### PR TITLE
Creates OfflineTransactionRefund event publishing

### PR DESCRIPTION
**Pull request by clarissalimab**
Part of [tix #2975](https://github.com/CommitChange/tix/issues/2975).

### CODE REVIEW

File: app/models/offline_transaction_refund.rb [hunk 1]
```diff
@@ -3,5 +3,86 @@
 # License: AGPL-3.0-or-later WITH WTO-AP-3.0-or-later
 # Full license explanation at https://github.com/houdiniproject/houdini/blob/master/LICENSE
 class OfflineTransactionRefund < ApplicationRecord
+	include Model::SubtransactionPaymentable
 	belongs_to :payment
+
+	delegate :gross_amount, :net_amount, :fee_total, :nonprofit, to: :payment
```
Comment (clarissalimab) [inline app/models/offline_transaction_refund.rb]: Is this the best way to refer to the nonprofit? I think so because it comes from payment, but maybe I'm wrong.

File: .rubocop.yml [hunk 2]
```diff
@@ -731,9 +731,12 @@ Layout/IndentationStyle:
 Layout/IndentationWidth:
   Width: 1
 
+Metrics/MethodLength:
+  Max: 25
+
```
Comment (clarissalimab) [inline .rubocop.yml:736]: The default is 10 but I think 25 is a more realistic length as it fits on the regular terminal size and we wouldn't have to ignore it.

File: app/models/offline_transaction_refund.rb (hunk 1 above)
Comment (wwahammy) [reply app/models/offline_transaction_refund.rb]: That's a way but it's not the best way.

A couple things to consider:
* `payment` is legacy. While the payment nonprofit there needs to match the one in the transaction, we're eventually going to use the one in the transaction only.
* `nonprofit` should be available via `has_one :nonprofit, through: :subtransaction_payment` which is included as a part of the [`Model::SubtransactionPaymentable` concern](https://github.com/wwahammy/houdini/blob/feature/event-publisher/offline-transaction-refund/app/models/concerns/model/subtransaction_paymentable.rb). By doing so, we should be able to run calls like `OfflineTransactionRefund.where(nonprofit: Nonprofit.find(1))` which is pretty slick!

Was there a specific reason that led to needing to add the delegate for `nonprofit` here instead?

File: .rubocop.yml (hunk 2 above)
Comment (wwahammy) [reply .rubocop.yml:736]: I support this change.

File: .rubocop.yml [hunk 3]
```diff
@@ -731,9 +731,12 @@ Layout/IndentationStyle:
...
 Metrics/BlockLength:
   Exclude:
-    - '**/*_spec.rb'
+    - 'spec/**/*.rb'
```
Comment (wwahammy) [inline .rubocop.yml:739]: Why the change here?

File: spec/support/examples/shared_offline_transactions_examples.rb [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+		it 'announces payment.deleted event' do
+			offline_transaction.publish_deleted
+
+			expect(event_publisher)
+				.to have_received(:announce)
+				.with(:payment_deleted, attributes)
+		end
+	end
+end
```
Comment (wwahammy) [inline spec/support/examples/shared_offline_transactions_examples.rb]: This is a good first pass! This stuff is not easy.

For the shared example, I want you to consider the situations where testing the results of OfflineTransactionRefund#publish_created might be useful: 

* Someone directly calls `OfflineTransactionRefund#publish_created`
* Someone calls `OfflineTransactionRefund.create` and we've added an `after_create` callback
* Someone calls an potential internal API inside the application like `ImaginaryOfflineTransactionProcessor.create_refund`
* Someone makes an HTTP call like `POST /api/transactions/v32134/refunds/`

In all of those cases, if we provide identical inputs, we know a matching `offline_transaction_refund.created` event will be published. Yes the HouIDs might be different, but we can use the `match_houid` method to match those. So the idea of a shared example is, given some sort of common interface in the spec description, the shared examples will run and we'll be able to verify whether these events were fired.

**Review by wwahammy — CHANGES_REQUESTED** [review]

File: .rubocop.yml (hunk 3 above)
Comment (clarissalimab) [reply .rubocop.yml:739]: Because of the shared example I created. I researched about it and I saw that it's common to ignore this BlockLength metric on tests, so I did it.

File: app/models/offline_transaction_refund.rb (hunk 1 above)
Comment (clarissalimab) [reply app/models/offline_transaction_refund.rb]: I'm not sure know why, when I was writing the tests I wasn't able to access `nonprofit` without doing it, but I just tested it again and it worked ò.ó

File: .rubocop.yml (hunk 3 above)
Comment (wwahammy) [reply .rubocop.yml:739]: Ah, I get it now. Let's go with that then.

File: app/models/offline_transaction_refund.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+# License: AGPL-3.0-or-later WITH WTO-AP-3.0-or-later
+# Full license explanation at https://github.com/houdiniproject/houdini/blob/master/LICENSE
+class OfflineTransactionRefund < ApplicationRecord
+	include Model::SubtransactionPaymentable
+	belongs_to :payment
+
+	delegate :gross_amount, :net_amount, :fee_total, :nonprofit, :supporter, to: :payment
```
Comment (clarissalimab) [inline app/models/offline_transaction_refund.rb]: With the code as it is, I can't access `nonprofit` and `supporter` without delegating to payment.

File: app/models/offline_transaction_refund.rb (hunk 1 above)
Comment (clarissalimab) [reply app/models/offline_transaction_refund.rb]: Update: https://github.com/wwahammy/houdini/pull/14/files#discussion_r650074781

**Comment by wwahammy** [comment]
Can you rebase on feature/event-publisher/offline-transaction-refund again?

File: spec/models/offline_transaction_refund_spec.rb [hunk 6]
```diff
@@ -5,5 +5,232 @@
...
+		end
+
+		it 'announces payment.deleted event' do
+			offline_transaction_refund.publish_deleted
+
+			expect(event_publisher)
+				.to have_received(:announce)
+				.with(:payment_deleted, attributes)
+		end
+	end
```
Comment (wwahammy) [inline spec/models/offline_transaction_refund_spec.rb]: These events need some work. 

We're trying to figure out whether the event being published looks like what we want it to look like. Mocking it out like this sort of defeats the purpose.

Please redo these and verify that the events look the way we want them to.

File: spec/models/offline_transaction_refund_spec.rb [hunk 7]
```diff
@@ -5,5 +5,232 @@
...
+			.and_return(event_publisher)
+		allow(event_publisher)
+			.to receive(:announce)
+			.with(any_args)
+		transaction
+	end
+
+	after do
+		Timecop.return
+	end
```
Comment (wwahammy) [inline spec/models/offline_transaction_refund_spec.rb]: Timecop should only be used as part of a block: i.e.

```ruby
Timecop.freeze(2020, 5,4) do 
 # The current time in this block is Time.new(2020, 5, 4)
end
```

Using Timecop outside a block risks the Time not being reset someplace and the frozen time leaking into a test we don't want it in.

If you want to have all of your examples into a `describe` or `context` block to use a specific frozen time, you put use an `around` block inside the describe or context:

```ruby
describe Foo do
 around(:each) do |example|
    Timecop.freeze(2020, 4, 5) do
      example.run
    end
  end

  it { # time is Time.new(2020, 4, 5)}

  it { # time is Time.new(2020, 4, 5)}

  it { # time is Time.new(2020, 4, 5)}
end
```

**Review by wwahammy — CHANGES_REQUESTED** [review]
Thanks for the fixes but the specs need some more work. We need to know that the events being published are correct.

**Comment by wwahammy** [comment]
@clarissalimab is this ready to review again?

**Comment by wwahammy** [comment]
@clarissalimab oops, I don't think it is :)

File: spec/models/offline_transaction_refund_spec.rb (hunk 7 above)
Comment (clarissalimab) [reply spec/models/offline_transaction_refund_spec.rb]: That's good to know :)

File: db/schema.rb [hunk 8]
```diff
@@ -1046,9 +1046,9 @@
   add_foreign_key "modern_campaign_gifts", "campaign_gifts"
   add_foreign_key "object_event_hook_configs", "nonprofits"
   add_foreign_key "offline_transaction_charges", "payments"
-  add_foreign_key "offline_transaction_refunds", "payments"
   add_foreign_key "recurrences", "recurring_donations"
   add_foreign_key "recurrences", "supporters"
+  add_foreign_key "offline_transaction_refunds", "payments"
```
Comment (wwahammy) [inline db/schema.rb:1051]: These should be alphabetical order: that's how the schema is created by Rails; if we don't the next time a migration is run, the schema will get reorganized and the changes will be a little confusing.

**Review by wwahammy — COMMENTED** [review]
This is really well done @clarissalimab. :+1:

There's one tweak to do but I'll actually do it myself :)
